REGRESSION(300618@main): `letter-spacing` transition no longer works

weinig opened the pull request:
#### 3a3971d9fc9294d0819735461d788028b1cb5c73
<pre>
REGRESSION(300618@main): `letter-spacing` transition no longer works
<a href="https://bugs.webkit.org/show_bug.cgi?id=304172">https://bugs.webkit.org/show_bug.cgi?id=304172</a>
<a href="https://rdar.apple.com/166530064">rdar://166530064</a>

Reviewed by Antoine Quint.

When splitting the computed letter-spacing and word-spacing values
out of FontCascade in 300618@main, I neglected to account for the
case where the the style builder is not used: animations. So, while
the computed letter-spacing and word-spacing values got updated
during animations (which is what most test cases look at because
it can be queried from script), the used letter-spacing and
word-spacing values on FontCascade where not getting synchronized.

To rectify this, I extracted the synchronization code into new
helpers on RenderStyleBase (with both updating, for animation, and
non-updating, for style building, variants), and introduced new
setters for the letter-spacing and word-spacing interpolator to
call that invokes the new synchronizers.

While thinking about this, I realized we had another related issue
with letter-spacing and word-spacing during animations but when
animating font-size. When animating font-size, we also need to ensure
that letter-spacing and word-spacing are synchronized, as they
might be using percentage values.

Tests: imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-font-size.html
       imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-letter-spacing.html
       imported/w3c/web-platform-tests/css/css-text/letter-spacing/reference/letter-spacing-animating-font-size-ref.html
       imported/w3c/web-platform-tests/css/css-text/letter-spacing/reference/letter-spacing-animating-letter-spacing-ref.html
       imported/w3c/web-platform-tests/css/css-text/word-spacing/reference/word-spacing-animating-font-size-ref.html
       imported/w3c/web-platform-tests/css/css-text/word-spacing/reference/word-spacing-animating-word-spacing-ref.html
       imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-font-size.html
       imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-word-spacing.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-font-size-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-font-size.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-letter-spacing-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-letter-spacing.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/reference/letter-spacing-animating-font-size-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/reference/letter-spacing-animating-letter-spacing-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/reference/word-spacing-animating-font-size-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/reference/word-spacing-animating-word-spacing-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-font-size-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-font-size.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-word-spacing-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-word-spacing.html: Added.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/style/RenderStyleBase.cpp:
* Source/WebCore/rendering/style/RenderStyleBase.h:
* Source/WebCore/rendering/style/RenderStylePropertiesSettersCustom.h:
* Source/WebCore/style/StyleBuilderState.cpp:

Canonical link: <a href="https://commits.webkit.org/304575@main">https://commits.webkit.org/304575@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c09c2d141431282067f633cb8597ef17a0cd79f3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135985 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8341 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47263 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143688 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/88946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137854 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9008 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8188 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103918 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/88946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138931 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6522 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121864 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84795 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/135333 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6211 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3863 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4290 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115476 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40055 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146439 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8026 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40623 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112273 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8045 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6736 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112666 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28591 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6127 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118165 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8074 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36234 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7795 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71631 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8016 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7876 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-font-size.html:
```diff
@@ -0,0 +1,34 @@
+<script>
+  const target = document.querySelector('#target');
+
+  let animation = target.animate(
```
`const` is fine here since you're not re-assigning that variable.

graouts commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-font-size.html:
```diff
@@ -0,0 +1,34 @@
+    [
+      { fontSize: '0px' },
+      { fontSize: '40px' }
+    ],
```
That's perfectly fine, but a more idiomatic way to write this would be:

```suggestion
    { fontSize: ['0px', '40px'] },
```

graouts commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-font-size.html:
```diff
@@ -0,0 +1,34 @@
+    ],
+    {
+      duration: 40
+    }
```
That too is perfectly fine, but if you only use a duration, you can just pass the value directly:

```suggestion
    40
```

graouts commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-letter-spacing.html:
```diff
@@ -0,0 +1,29 @@
+    {
+      duration: 40
+    }
+  );
```
Replicating my earlier comments, you could write this as:

```suggestion
  const animation = target.animate({ letterSpacing: ['0px', '40px'] }, 40);
```

graouts commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-font-size.html:
```diff
@@ -0,0 +1,34 @@
+    {
+      duration: 40
+    }
+  );
```
```suggestion
  const animation = target.animate({ fontSize: ['0px', '40px'] }, 40);
```

graouts commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-word-spacing.html:
```diff
@@ -0,0 +1,29 @@
+    {
+      duration: 40
+    }
+  );
```
```suggestion
  const animation = target.animate({ wordSpacing: ['0px', '40px'] }, 40);
```

graouts commented on Source/WebCore/rendering/style/RenderStyleBase.cpp:
```diff
@@ -177,7 +177,6 @@ const FontMetrics& RenderStyleBase::metricsOfPrimaryFont() const
     return m_inheritedData->fontData->fontCascade.metricsOfPrimaryFont();
 }
 
-
```
Nice.

graouts commented on Source/WebCore/rendering/style/RenderStyleBase.cpp:
```diff
@@ -219,6 +218,88 @@ void RenderStyleBase::setSpecifiedLineHeight(Style::LineHeight&& lineHeight)
+        if (oldShouldDisableLigatures != newShouldDisableLigatures) {
+            auto newFontDescription = oldFontDescription;
+            newFontDescription.setShouldDisableLigaturesForSpacing(newShouldDisableLigatures);
+            setFontDescriptionWithoutUpdate(WTFMove(newFontDescription));
```
This seems to be the only line that is different compared to `synchronizeLetterSpacingWithFontCascade()`. Perhaps we could find a way to have these two sharing their code passing only a flag that indicates whether to update?

weinig replied:
I left this alone for now, since it is not that much duplication and I want to mess around with this more soon.

weinig committed:
REGRESSION(300618@main): `letter-spacing` transition no longer works
https://bugs.webkit.org/show_bug.cgi?id=304172
rdar://166530064

Reviewed by Antoine Quint.

When splitting the computed letter-spacing and word-spacing values
out of FontCascade in 300618@main, I neglected to account for the
case where the the style builder is not used: animations. So, while
the computed letter-spacing and word-spacing values got updated
during animations (which is what most test cases look at because
it can be queried from script), the used letter-spacing and
word-spacing values on FontCascade where not getting synchronized.

To rectify this, I extracted the synchronization code into new
helpers on RenderStyleBase (with both updating, for animation, and
non-updating, for style building, variants), and introduced new
setters for the letter-spacing and word-spacing interpolator to
call that invokes the new synchronizers.

While thinking about this, I realized we had another related issue
with letter-spacing and word-spacing during animations but when
animating font-size. When animating font-size, we also need to ensure
that letter-spacing and word-spacing are synchronized, as they
might be using percentage values.

Tests: imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-font-size.html
       imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-letter-spacing.html
       imported/w3c/web-platform-tests/css/css-text/letter-spacing/reference/letter-spacing-animating-font-size-ref.html
       imported/w3c/web-platform-tests/css/css-text/letter-spacing/reference/letter-spacing-animating-letter-spacing-ref.html
       imported/w3c/web-platform-tests/css/css-text/word-spacing/reference/word-spacing-animating-font-size-ref.html
       imported/w3c/web-platform-tests/css/css-text/word-spacing/reference/word-spacing-animating-word-spacing-ref.html
       imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-font-size.html
       imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-word-spacing.html
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-font-size-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-font-size.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-letter-spacing-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/letter-spacing-animating-letter-spacing.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/reference/letter-spacing-animating-font-size-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/letter-spacing/reference/letter-spacing-animating-letter-spacing-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/reference/word-spacing-animating-font-size-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/reference/word-spacing-animating-word-spacing-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-font-size-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-font-size.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-word-spacing-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/word-spacing/word-spacing-animating-word-spacing.html: Added.
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/rendering/style/RenderStyleBase.cpp:
* Source/WebCore/rendering/style/RenderStyleBase.h:
* Source/WebCore/rendering/style/RenderStylePropertiesSettersCustom.h:
* Source/WebCore/style/StyleBuilderState.cpp:

Canonical link: https://commits.webkit.org/304575@main
